power-sequencer/mihawk-cpld.cpp

```diff
@@ -0,0 +1,478 @@
+    }
+
+    // return the i2c-read data
+    return statusValue_2;
+}
+
+// Check PSU_present_pin.
+bool MihawkCPLD::checkPSUdevice()
+{
+    bool result = 0;
+    int bin[32], num;
+    int i = 0, length = 0;
+    std::string i2cBus = "/dev/i2c-" + std::to_string(busId);
+
+    // open i2c device(CPLD-PSU-register table)
+    int fd = open(i2cBus.c_str(), O_RDWR | O_CLOEXEC);
+    if (fd < 0)
+    {
+        std::cerr << "Unable to open i2c device \n";
+    }
+
+    // set i2c slave address
+    if (ioctl(fd, I2C_SLAVE_FORCE, slaveAddr) < 0)
+    {
+        std::cerr << "Unable to set device address \n";
+        close(fd);
+    }
+
+    // check whether support i2c function
+    unsigned long funcs = 0;
+    if (ioctl(fd, I2C_FUNCS, &funcs) < 0)
+    {
+        std::cerr << "Not support I2C_FUNCS \n";
+        close(fd);
+    }
+
+    // check whether support i2c-read function
+    if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE_DATA))
+    {
+        std::cerr << "Not support I2C_FUNC_SMBUS_READ_BYTE_DATA \n";
+        close(fd);
+    }
+
+    int statusValue;
+    unsigned int statusReg = StatusReg;
+
+    statusValue = i2c_smbus_read_byte_data(fd, statusReg);
+    close(fd);
+
+    if (statusValue < 0)
+    {
+        std::cerr << "i2c_smbus_read_byte_data failed \n";
+    }
+
+    // Transfer decimal_value to binary_value
+    num = statusValue;
+    do
+    {
+        bin[i] = num % 2;
+        num = num / 2;
+        i++;
+        length++;
+    } while (num != 1);
+    bin[length] = num;
+    length++;
+
+    if (bin[3] == 1 && bin[4] == 1)
+    {
+        // If power_on-interrupt-bit is read as 1,
+        // switch on the flag.
+        result = 1;
+    }
+    else
+    {
+        result = 0;
+    }
+
+    return result;
+}
+
+// Check for PoweronFault
+bool MihawkCPLD::checkPoweronFault()
+{
+    bool result_1 = 0;
+    int bin1[32], num1;
+    int i = 0, length = 0;
+    std::string i2cBus = "/dev/i2c-" + std::to_string(busId);
+
+    // open i2c device(CPLD-PSU-register table)
+    int fd = open(i2cBus.c_str(), O_RDWR | O_CLOEXEC);
+    if (fd < 0)
+    {
+        std::cerr << "Unable to open i2c device \n";
+    }
+
+    // set i2c slave address
+    if (ioctl(fd, I2C_SLAVE_FORCE, slaveAddr) < 0)
+    {
+        std::cerr << "Unable to set device address \n";
+        close(fd);
+    }
+
+    // check whether support i2c function
+    unsigned long funcs = 0;
+    if (ioctl(fd, I2C_FUNCS, &funcs) < 0)
+    {
+        std::cerr << "Not support I2C_FUNCS \n";
+        close(fd);
+    }
+
+    // check whether support i2c-read function
+    if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE_DATA))
+    {
+        std::cerr << "Not support I2C_FUNC_SMBUS_READ_BYTE_DATA \n";
+        close(fd);
+    }
+
+    int statusValue_1;
+    unsigned int statusReg_1 = StatusReg_1;
+
+    statusValue_1 = i2c_smbus_read_byte_data(fd, statusReg_1);
+    close(fd);
+
+    if (statusValue_1 < 0)
+    {
+        std::cerr << "i2c_smbus_read_byte_data failed \n";
+    }
```

Comment:
same comment about this giving back a bad result if the above i2c_smbus... function failed.

Reply:
This part is my negligence, I will fix it again.